File: 3559.number-of-ways-to-assign-edge-weights-ii.go
#
# @lc app=leetcode id=3559 lang=golang
#
# [3559] Number of Ways to Assign Edge Weights II
#

# @lc code=start
func assignEdgeWeights(edges [][]int, queries [][]int) []int {
	const MOD int = 1000000007
	const LOG int = 18
	n := len(edges) + 1
	adj := make([][]int, n+1)
	for _, e := range edges {
		adj[e[0]] = append(adj[e[0]], e[1])
		adj[e[1]] = append(adj[e[1]], e[0])
	}
	parent := make([][]int, n+1)
	for i := 1; i <= n; i++ {
		parent[i] = make([]int, LOG)
	}
	depth := make([]int, n+1)
	visited := make([]bool, n+1)

	// BFS to compute depth and parent[0]
	q := make([]int, 0, n+1)
	front := 0
	q = append(q, 1)
	visited[1] = true
	parent[1][0] = 1
	depth[1] = 0
	for front < len(q) {
		u := q[front]
		front++
		for _, v := range adj[u] {
			if !visited[v] {
				visited[v] = true
				parent[v][0] = u
				depth[v] = depth[u] + 1
				q = append(q, v)
			}
		}
	}

	// Build binary lifting table
	for j := 1; j < LOG; j++ {
		for i := 1; i <= n; i++ {
			parent[i][j] = parent[parent[i][j-1]][j-1]
		}
	}

	// Modular exponentiation
	modpow := func(base, exp int) int {
		res := 1
		base %= MOD
		for exp > 0 {
			if exp&1 == 1 {
				res = int(int64(res) * int64(base) % int64(MOD))
			}
			base = int(int64(base) * int64(base) % int64(MOD))
			exp >>= 1
		}
		return res
	}

	// LCA function
	getLCA := func(u, v int) int {
		if depth[u] > depth[v] {
			u, v = v, u
		}
		diff := depth[v] - depth[u]
		for j := 0; j < LOG; j++ {
			if diff&(1<<j) != 0 {
				v = parent[v][j]
			}
		}
		if u == v {
			return u
		}
		for j := LOG - 1; j >= 0; j-- {
			if parent[u][j] != parent[v][j] {
				u = parent[u][j]
				v = parent[v][j]
			}
		}
		return parent[u][0]
	}

	answer := make([]int, len(queries))
	for i := range queries {
		u := queries[i][0]
		v := queries[i][1]
		if u == v {
			answer[i] = 0
			continue
		}
		lca_node := getLCA(u, v)
		path_len := depth[u] + depth[v] - 2*depth[lca_node]
		answer[i] = modpow(2, path_len-1)
	}
	return answer
}
# @lc code=end

File: 3559.number-of-ways-to-assign-edge-weights-ii.java
#
# @lc app=leetcode id=3559 lang=java
#
# [3559] Number of Ways to Assign Edge Weights II
#

# @lc code=start
class Solution {
    public int[] assignEdgeWeights(int[][] edges, int[][] queries) {
        int n = edges.length + 1;
        @SuppressWarnings("unchecked")
        List<Integer>[] adj = new ArrayList[n + 1];
        for (int i = 0; i <= n; i++) {
            adj[i] = new ArrayList<>();
        }
        for (int[] e : edges) {
            int u = e[0], v = e[1];
            adj[u].add(v);
            adj[v].add(u);
        }
        final long MOD = 1000000007L;
        final int LOG = 18;
        int[][] par = new int[LOG][n + 1];
        int[] dep = new int[n + 1];
        boolean[] vis = new boolean[n + 1];
        Queue<Integer> q = new LinkedList<>();
        q.offer(1);
        vis[1] = true;
        dep[1] = 0;
        par[0][1] = -1;
        while (!q.isEmpty()) {
            int u = q.poll();
            for (int v : adj[u]) {
                if (!vis[v]) {
                    vis[v] = true;
                    dep[v] = dep[u] + 1;
                    par[0][v] = u;
                    q.offer(v);
                }
            }
        }
        for (int k = 1; k < LOG; k++) {
            for (int i = 1; i <= n; i++) {
                if (par[k - 1][i] != -1) {
                    par[k][i] = par[k - 1][par[k - 1][i]];
                } else {
                    par[k][i] = -1;
                }
            }
        }
        int[] ans = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int u = queries[i][0];
            int v = queries[i][1];
            if (u == v) {
                ans[i] = 0;
                continue;
            }
            int uu = u;
            int vv = v;
            if (dep[uu] > dep[vv]) {
                int temp = uu;
                uu = vv;
                vv = temp;
            }
            int diff = dep[vv] - dep[uu];
            for (int k = 0; k < LOG; k++) {
                if ((diff & (1 << k)) != 0) {
                    vv = par[k][vv];
                }
            }
            if (uu == vv) {
                int l = uu;
                int dist = dep[u] + dep[v] - 2 * dep[l];
                ans[i] = (int) modpow(2L, (long)dist - 1L, MOD);
                continue;
            }
            for (int k = LOG - 1; k >= 0; k--) {
                if (par[k][uu] != par[k][vv]) {
                    uu = par[k][uu];
                    vv = par[k][vv];
                }
            }
            int l = par[0][uu];
            int dist = dep[u] + dep[v] - 2 * dep[l];
            ans[i] = (int) modpow(2L, (long)dist - 1L, MOD);
        }
        return ans;
    }

    private static long modpow(long b, long e, long mod) {
        long res = 1;
        b %= mod;
        while (e > 0) {
            if ((e & 1) == 1) {
                res = res * b % mod;
            }
            b = b * b % mod;
            e >>= 1;
        }
        return res;
    }
}
# @lc code=end